From 79f7511d8b8d92b940e227f056f3e4e0947c0a20 Mon Sep 17 00:00:00 2001
From: Dylan Hung <dylan_hung@aspeedtech.com>
Date: Fri, 23 Sep 2022 17:09:23 +0800
Subject: [PATCH 21/30] i3c: aspeed: reset the IBI queue if unregistered IBI
 source

If the IBI source is not registered, ignore the IBI request by resetting
the IBI queue.

Signed-off-by: Dylan Hung <dylan_hung@aspeedtech.com>
Change-Id: I661b7f8642f3717ff50d2f6e6174a3f4c6618ec1
---
 drivers/i3c/i3c_aspeed.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/i3c/i3c_aspeed.c b/drivers/i3c/i3c_aspeed.c
index 5e8c621c46..5d767442d6 100644
--- a/drivers/i3c/i3c_aspeed.c
+++ b/drivers/i3c/i3c_aspeed.c
@@ -763,6 +763,11 @@ static void i3c_aspeed_master_rx_ibi(struct i3c_aspeed_obj *obj)
 		}
 
 		pos = i3c_aspeed_get_pos(obj, ibi_status.id >> 1);
+		if (pos < 0) {
+			LOG_ERR("unregistered IBI source: 0x%x\n", ibi_status.id >> 1);
+			i3c_register->reset_ctrl.fields.ibi_queue_reset = 1;
+			continue;
+		}
 
 		i3cdev = obj->dev_descs[pos];
 		priv = DESC_PRIV(i3cdev);
-- 
2.24.1

